### PR TITLE
kpatch-build: support CONFIG_LTO_CLANG_THIN

### PR DESCRIPTION
**Pull request by liu-song-6**
Support CONFIG_LTO_CLANG_THIN with ld.lld --lto-obj-path option.

With CONFIG_LTO_CLANG_THIN, .o files are LLVM IR binary, so CDO doesn't work on .o file. To solve this issue, we CDO the thinlto files generated by the --lto-obj-path option. Clang LTO generates the thinlto files after cross file inline, so they are good candidates for CDO. See [1] for more discussions about this.

To achieve this, we need:

  1. kpatch-build to update kernel Makefile(s) so it generates thinlto files;
  2. kpatch-build and kpatch-cc to save the thinlto file properly;
  3. kpatch-build to feed these thinlto files to CDO;
  4. The user need to supply vmlinux.o, from which we generate the symtab file. We need this because GLOBAL symbols may be marked as LOCAL in LTO vmlinux;
  4. A small workaround in CDO that ignores changes in init.text for vmlinux.o.thinlto.*

[1] https://github.com/dynup/kpatch/issues/1320

Signed-off-by: Song Liu <song@kernel.org>

### CODE REVIEW

File: kpatch-build/kpatch-build [hunk 1]
```diff
@@ -1054,6 +1060,35 @@ if [[ -n "$CONFIG_DEBUG_INFO_BTF" ]]; then
...
+	[[ -e "$VMLINUX" ]] || die "For kernel with CONFIG_LTO_CLANG, please supply vmlinux.o via -v|--vmlinux option."
+
+	# This is a heuristic: use -x to check vmlinux vs. vmlinux.o
+	[[ -x "$VMLINUX" ]] && die "For kernel with CONFIG_LTO_CLANG, please supply vmlinux.o instead of vmlinux via -v|--vmlinux option."
+
+	cp -f "$KERNEL_SRCDIR/Makefile" "$TEMPDIR/Makefile" || die
+	sed -i "s/--thinlto-cache-dir=\$(extmod_prefix).thinlto-cache/--lto-obj-path=vmlinux.o.thinlto.o/g" "$KERNEL_SRCDIR"/Makefile
+
+	cp -f "$KERNEL_SRCDIR/scripts/Makefile.build" "$TEMPDIR/Makefile.build" || die
+	sed -i "s/\$(ld_flags)/\$(ld_flags) --lto-obj-path=\$@.thinlto.o/g" "$KERNEL_SRCDIR"/scripts/Makefile.build
```
Comment (jpoimboe) [inline kpatch-build/kpatch-build:1087]: These changes to Makefile and Makefile.build need short comments describing what they're doing and why

File: kpatch-build/kpatch-cc [hunk 2]
```diff
@@ -79,6 +79,24 @@ elif [[ "$TOOLCHAINCMD" =~ ^(.*-)?ld || "$TOOLCHAINCMD" =~ ^(.*-)?ld.lld ]] ; th
 					args+=(--warn-unresolved-symbols)
 					break
 					;;
+				*/.tmp_*.o)
+					# .tmp_*.o is used for single file modules.
+					# See "cmd_ld_single_m" in scripts/Makefile.build.
+					if [[ $KPATCH_CC_HANDLE_LTO -ne 0 ]] ; then
```
Comment (jpoimboe) [inline kpatch-build/kpatch-cc:85]: I think "[[ -n $KPATCH_CC_LTO -eq 1 ]]" would be clearer since we never actually initialize the variable to 0 (and also it avoids the double negative check)

File: kpatch-build/kpatch-cc [hunk 3]
```diff
@@ -79,6 +79,24 @@ elif [[ "$TOOLCHAINCMD" =~ ^(.*-)?ld || "$TOOLCHAINCMD" =~ ^(.*-)?ld.lld ]] ; th
...
+						mkdir -p "$KPATCH_GCC_TEMPDIR/orig/$(dirname "$relobj")"
+						cp "${obj/.tmp_/}".thinlto.o* "$KPATCH_GCC_TEMPDIR/orig/$(dirname "$relobj")"
+						echo "${obj/.tmp_/}".thinlto.o* >> "$KPATCH_GCC_TEMPDIR/thinlto_objs"
+					fi
+					break
+					;;
+				*.o)
+					if [[ $KPATCH_CC_HANDLE_LTO -ne 0 ]] ; then
```
Comment (jpoimboe) [inline kpatch-build/kpatch-cc:93]: same

File: kpatch-build/kpatch-build [hunk 4]
```diff
@@ -1238,6 +1284,19 @@ for i in $FILES; do
...
+				KOBJFILE_PATH="$VMLINUX"
+				SYMTAB="${TEMPDIR}/${KOBJFILE_NAME}.symtab"
+				SYMVERS_FILE="$BUILDDIR/Module.symvers"
+			else
+				KOBJFILE_NAME=$(basename "${KOBJFILE%.ko}")
+				KOBJFILE_NAME="${KOBJFILE_NAME//-/_}"
+				KOBJFILE_PATH="${TEMPDIR}/module/$KOBJFILE.ko"
+				SYMTAB="${KOBJFILE_PATH}.symtab"
+				SYMVERS_FILE="$BUILDDIR/Module.symvers"
+			fi
```
Comment (jpoimboe) [inline kpatch-build/kpatch-build:1299]: This change is almost completely copy-pasted from existing code and could surely be simplified?

File: kpatch-build/kpatch-build [hunk 5]
```diff
@@ -1229,7 +1271,11 @@ for i in $FILES; do
 
 	mkdir -p "output/$(dirname "$i")"
 	cd "$BUILDDIR" || die
-	find_kobj "$i"
+	if [[ -z "$CONFIG_LTO_CLANG" ]] ; then
+		find_kobj "$i"
+	else
+		KOBJFILE=${i/.o.thinlto.o*/}
+	fi
```
Comment (jpoimboe) [inline kpatch-build/kpatch-build:1278]: Please do the `[[ -n "$CONFIG_LTO_CLANG ]]` case first

File: kpatch-build/kpatch-build [hunk 6]
```diff
@@ -1107,6 +1142,7 @@ declare -a MAKEVARS
 if [[ -n "$CONFIG_CC_IS_CLANG" ]]; then
 	MAKEVARS+=("CC=${KPATCH_CC_PREFIX}${CLANG}")
 	MAKEVARS+=("HOSTCC=clang")
+	MAKEVARS+=("LLVM=1")
```
Comment (jpoimboe) [inline kpatch-build/kpatch-build:1145]: This is only a one-line change but it has a big impact and belongs in a separate commit with proper justification.  Also I wonder how it interacts with the CONFIG_LD_IS_LLD check.

File: kpatch-build/kpatch-build [hunk 7]
```diff
@@ -1054,6 +1060,35 @@ if [[ -n "$CONFIG_DEBUG_INFO_BTF" ]]; then
...
+	# This is a heuristic: use -x to check vmlinux vs. vmlinux.o
+	[[ -x "$VMLINUX" ]] && die "For kernel with CONFIG_LTO_CLANG, please supply vmlinux.o instead of vmlinux via -v|--vmlinux option."
+
+	cp -f "$KERNEL_SRCDIR/Makefile" "$TEMPDIR/Makefile" || die
+	sed -i "s/--thinlto-cache-dir=\$(extmod_prefix).thinlto-cache/--lto-obj-path=vmlinux.o.thinlto.o/g" "$KERNEL_SRCDIR"/Makefile
+
+	cp -f "$KERNEL_SRCDIR/scripts/Makefile.build" "$TEMPDIR/Makefile.build" || die
+	sed -i "s/\$(ld_flags)/\$(ld_flags) --lto-obj-path=\$@.thinlto.o/g" "$KERNEL_SRCDIR"/scripts/Makefile.build
+
+	export KPATCH_CC_HANDLE_LTO=1
```
Comment (jpoimboe) [inline kpatch-build/kpatch-build:1089]: I don't think "HANDLE" adds anything, can we just call it KPATCH_CC_LTO?

File: kpatch-build/kpatch-build [hunk 8]
```diff
@@ -1174,9 +1210,14 @@ if [[ -n "$CONFIG_MODVERSIONS" ]]; then
 	trace_on
 fi
 
+if [[ -n "$CONFIG_LTO_CLANG" ]]; then
+	DIFF_OBJS="$TEMPDIR/thinlto_objs"
+else
+	DIFF_OBJS="$TEMPDIR/changed_objs"
+fi
```
Comment (jpoimboe) [inline kpatch-build/kpatch-build:1217]: Why does lto need its own special file name?

File: kpatch-build/kpatch-build [hunk 9]
```diff
@@ -1251,11 +1310,21 @@ for i in $FILES; do
 			SYMVERS_FILE="$BUILDDIR/Module.symvers"
 		fi
 
-		"$READELF" -s --wide "$KOBJFILE_PATH" > "$SYMTAB"
+		# With CONFIG_LTO_CLANG, multiple .thinlto files share a
+		# symtab file. Only generate the symtab file once.
+		[[ -e "$SYMTAB" ]] || "$READELF" -s --wide "$KOBJFILE_PATH" > "$SYMTAB"
```
Comment (jpoimboe) [inline kpatch-build/kpatch-build:1315]: For more readability use the long arg: "--symbols"

File: kpatch-build/create-diff-object.c [hunk 10]
```diff
@@ -2825,6 +2825,21 @@ static void kpatch_mark_ignored_sections(struct kpatch_elf *kelf)
...
+		 * while the original function is very similar, like:
+		 * __initstub__kmod_syscall__728_5324_bpf_syscall_sysctl_init7
+		 *
+		 * We don't have very good solution for it yet. To workaround
+		 * the issue, it seems safe to ignore .init.text changes in
+		 * vmlinux.o. This will skip such new functions.
+		 */
+		if (!strncmp(childobj, "vmlinux.o", 9) &&
+		    !strncmp(sec->name, ".init.text", 10))
+			sec->ignore = 1;
```
Comment (jpoimboe) [inline kpatch-build/create-diff-object.c:2842]: Also this (or whatever the change eventually is) should probably be in its own commit.

File: kpatch-build/create-diff-object.c [hunk 11]
```diff
@@ -2825,6 +2825,21 @@ static void kpatch_mark_ignored_sections(struct kpatch_elf *kelf)
 		    !strncmp(sec->name, ".llvm_addrsig", 13) ||
 		    !strncmp(sec->name, ".llvm.", 6))
 			sec->ignore = 1;
+
+		/*
+		 * With CONFIG_LTO_CLANG, we see some weird new functions,
+		 * such as:
+		 * __initstub__kmod_syscall__728_5326_bpf_syscall_sysctl_init7
+		 * while the original function is very similar, like:
+		 * __initstub__kmod_syscall__728_5324_bpf_syscall_sysctl_init7
```
[the remark this answers is not in the document]
Comment (liu-song-6) [reply kpatch-build/create-diff-object.c:2834]: I played with kpatch_mangled_strcmp() a little bit, but didn't solve all the problem. Let me dig more into it.

File: kpatch-build/kpatch-build [hunk 12]
```diff
@@ -179,6 +179,12 @@ cleanup() {
 	# restore original Makefile.modfinal if we updated it for the build
 	[[ -e "$TEMPDIR/Makefile.modfinal" ]] && mv -f "$TEMPDIR/Makefile.modfinal" "$KERNEL_SRCDIR/scripts"
 
+	# restore original Makefile.build if we updated it for the build
+	[[ -e "$TEMPDIR/Makefile.build" ]] && mv -f "$TEMPDIR/Makefile.build" "$KERNEL_SRCDIR/scripts"
+
+	# restore Makefile if we updated it for the build
+	[[ -e "$TEMPDIR/Makefile" ]] && mv -f "$TEMPDIR/Makefile" "$KERNEL_SRCDIR"
+
```
[the remark this answers is not in the document]
Comment (liu-song-6) [reply kpatch-build/kpatch-build:187]: Fixed.

File: kpatch-build/kpatch-build [hunk 13]
```diff
@@ -1054,6 +1060,35 @@ if [[ -n "$CONFIG_DEBUG_INFO_BTF" ]]; then
...
+	#
+	# readelf -s vmlinux  | grep perf_event_bpf_event
+	#
+	# non-LTO:
+	#  124946: ffffffff81287070   858 FUNC    GLOBAL DEFAULT    1 perf_event_bpf_event
+	#
+	# LTO:
+	#  122028: ffffffff81298cd0   861 FUNC    LOCAL  HIDDEN     1 perf_event_bpf_event
+	#
+	# To work around this, use vmlinux.o instead.
```
[the remark this answers is not in the document]
Comment (liu-song-6) [reply kpatch-build/kpatch-build:1077]: Sounds great! Fixed it.

File: kpatch-build/kpatch-build (hunk 7 above)
Comment (liu-song-6) [reply kpatch-build/kpatch-build:1089]: Sure. Fixed.

File: kpatch-build/kpatch-build (hunk 6 above)
Comment (liu-song-6) [reply kpatch-build/kpatch-build:1145]: It appears to me that I need to pass in LLVM=1 to the make command. Otherwise, the make will fail. I have the same change for PGO support.

File: kpatch-build/kpatch-build (hunk 8 above)
Comment (liu-song-6) [reply kpatch-build/kpatch-build:1217]: Initially, I thought we may still use changed_objs for something else. But I think we didn't use it at the moment. 

OTOH, "changed_objs" is not accurate for thin LTO, as not all these files are changed.

File: kpatch-build/kpatch-build (hunk 9 above)
Comment (liu-song-6) [reply kpatch-build/kpatch-build:1315]: Fixed.

File: kpatch-build/kpatch-build (hunk 1 above)
Comment (liu-song-6) [reply kpatch-build/kpatch-build:1071]: Yeah, updating kpatch_mangled_strcmp() solves the problem.

File: kpatch-build/kpatch-build (hunk 6 above)
Comment (liu-song-6) [reply kpatch-build/kpatch-build]: It is possible for the user to specify CC=clang and LD=ld.lld separately. And LLVM=1 will enable a set of utilities. From llvm.rst:
```
LLVM has substitutes for GNU binutils utilities. They can be enabled individually.
The full list of supported make variables::
        make CC=clang LD=ld.lld AR=llvm-ar NM=llvm-nm STRIP=llvm-strip \
          OBJCOPY=llvm-objcopy OBJDUMP=llvm-objdump READELF=llvm-readelf \
          HOSTCC=clang HOSTCXX=clang++ HOSTAR=llvm-ar HOSTLD=ld.lld
To simplify the above command, Kbuild supports the ``LLVM`` variable::
        make LLVM=1
```
I changed the code to use CONFIG_AS_IS_LLVM to detect the use of LLVM=1. I think this should cover most use cases.

**Comment by liu-song-6** [comment]
@jpoimboe Does this version look reasonable?

File: kpatch-build/kpatch-build [hunk 14]
```diff
@@ -1118,6 +1137,11 @@ else
 	MAKEVARS+=("LD=${KPATCH_CC_PREFIX}${LD}")
 fi
 
+# It is common for the user to enable all LLVM utilities with LLVM=1.
+# Detech such case with CONFIG_AS_IS_LLVM flag.
+if [[ -n "$CONFIG_AS_IS_LLVM" ]]; then
+	MAKEVARS+=("LLVM=1")
+fi
```
[the remark this answers is not in the document]
Comment (liu-song-6) [reply kpatch-build/kpatch-build:1144]: Yeah, I couldn't think of a better test for LLVM=1.

File: kpatch-build/kpatch-build [hunk 15]
```diff
@@ -1054,6 +1054,25 @@ if [[ -n "$CONFIG_DEBUG_INFO_BTF" ]]; then
 	fi
 fi
 
+if [[ -n "$CONFIG_LTO_CLANG" ]]; then
+	[[ -n "$CONFIG_LTO_CLANG_THIN" ]] || die "Non-thin LTO is not supported. Please enable CONFIG_LTO_CLANG_THIN"
+
+	# With LTO, many GLOBAL symbols are converted to LOCAL in vmlinux,
+	# which confuses the klp reloc logic.  Require vmlinux.o for LTO.
+	[[ -e "$VMLINUX" ]] || die "For kernel with CONFIG_LTO_CLANG, please supply vmlinux.o via -v|--vmlinux option."
```
[the remark this answers is not in the document]
Comment (liu-song-6) [reply kpatch-build/kpatch-build:1062]: Will remove.

File: kpatch-build/create-diff-object.c [hunk 16]
```diff
@@ -430,6 +430,30 @@ static int __kpatch_unique_id_strcmp(char *s1, char *s2)
...
+		if (*s1 == '\0' || *s2 == '\0')
+			break;
+
+		s1++;
+		s2++;
+	}
+
+	return *s1 == *s2;
+}
+
```
[the remark this answers is not in the document]
Comment (liu-song-6) [reply kpatch-build/create-diff-object.c:456]: Hmm.. I am also confused with 1/0 return value. Let me double check..

[the remark this answers is not in the document]
Comment (liu-song-6) [reply kpatch-build/create-diff-object.c:456]: > Here's a simplified version (untested)
> 
> ```
> static int __kpatch_skip_digits_strcmp(char *s1, char *s2)
> {
> 	while (*s1 == *s2) {
> 		if (!*s1)
> 			return 0;
> 		s1++;
> 		s2++;
> 
> 		while (isdigit(*s1))
> 			s1++;
> 		while (isdigit(*s2))
> 			s2++;
> 	}
> 
> 	return 1;
> }
> ```

This is really clean. Thanks!

**Comment by liu-song-6** [comment]
@jpoimboe Could you please share your comments on the latest version? Thanks in advance!

**Comment by jpoimboe** [comment]
I tried out the patches against a 6.2 kernel and got the following errors:
```
Extracting new and modified ELF sections
Binary files orig/vmlinux.o.thinlto.o2 and patched/vmlinux.o.thinlto.o2 differ
Binary files orig/vmlinux.o.thinlto.o223 and patched/vmlinux.o.thinlto.o223 differ
create-diff-object: ERROR: vmlinux.o.thinlto.o223: kpatch_bundle_symbols: 244: symbol trace_event_raw_event_task_newtask at offset 16 within section .text.trace_event_raw_event_task_newtask, expected 0
Binary files orig/vmlinux.o.thinlto.o227 and patched/vmlinux.o.thinlto.o227 differ
create-diff-object: ERROR: vmlinux.o.thinlto.o227: kpatch_bundle_symbols: 244: symbol delayed_put_task_struct at offset 16 within section .text.delayed_put_task_struct, expected 0
Binary files orig/vmlinux.o.thinlto.o232 and patched/vmlinux.o.thinlto.o232 differ
create-diff-object: ERROR: vmlinux.o.thinlto.o232: kpatch_bundle_symbols: 244: symbol __ptrace_may_access at offset 16 within section .text.__ptrace_may_access, expected 0
Binary files orig/vmlinux.o.thinlto.o234 and patched/vmlinux.o.thinlto.o234 differ
create-diff-object: ERROR: vmlinux.o.thinlto.o234: kpatch_bundle_symbols: 244: symbol trace_event_raw_event_signal_generate at offset 16 within section .text.trace_event_raw_event_signal_generate, expected 0
Binary files orig/vmlinux.o.thinlto.o307 and patched/vmlinux.o.thinlto.o307 differ
create-diff-object: ERROR: vmlinux.o.thinlto.o307: kpatch_bundle_symbols: 244: symbol __klp_shadow_get_or_alloc at offset 16 within section .text.__klp_shadow_get_or_alloc, expected 0
Binary files orig/vmlinux.o.thinlto.o469 and patched/vmlinux.o.thinlto.o469 differ
create-diff-object: ERROR: vmlinux.o.thinlto.o469: kpatch_bundle_symbols: 244: symbol trace_event_raw_event_vm_unmapped_area at offset 16 within section .text.trace_event_raw_event_vm_unmapped_area, expected 0
Binary files orig/vmlinux.o.thinlto.o510 and patched/vmlinux.o.thinlto.o510 differ
create-diff-object: ERROR: vmlinux.o.thinlto.o510: kpatch_bundle_symbols: 244: symbol get_arg_page at offset 16 within section .text.get_arg_page, expected 0
Binary files orig/vmlinux.o.thinlto.o593 and patched/vmlinux.o.thinlto.o593 differ
create-diff-object: ERROR: vmlinux.o.thinlto.o593: kpatch_bundle_symbols: 244: symbol do_task_stat at offset 16 within section .text.do_task_stat, expected 0
Binary files orig/vmlinux.o.thinlto.o809 and patched/vmlinux.o.thinlto.o809 differ
create-diff-object: ERROR: vmlinux.o.thinlto.o809: kpatch_bundle_symbols: 244: symbol selinux_set_mnt_opts at offset 16 within section .text.selinux_set_mnt_opts, expected 0
ERROR: 9 error(s) encountered. Check /home/jpoimboe/.kpatch/build.log for more details.
```
The cause for the errors is the new CONFIG_CALL_DEPTH_TRACKING feature which enables CONFIG_PREFIX_SYMBOLS, which relies on objtool to create a `__pfx_<func>` symbol for the 16 bytes of NOPS preceding each function.

The problem is that objtool didn't run on the vmlinux.o.thinlto.o* objects.

Objtool has become an integral part of the x86 toolchain and removing it from patch module objects could create all kinds of surprises.  So I'm thinking this approach may not work and we may have to add vmlinux.o support to CDO after all.

**Comment by jpoimboe** [comment]
> How should we enable CDOing vmlinux.o? @jpoimboe IIRC, you plan to work on this soon?

I could certainly do it, though I'm probably not free for at least a few weeks.

**Comment by liu-song-6** [comment]
> > How should we enable CDOing vmlinux.o? @jpoimboe IIRC, you plan to work on this soon?
> 
> I could certainly do it, though I'm probably not free for at least a few weeks.

At the moment, I am hoping this version will work for our 5.19 based kernel. If it doesn't work, we would probably need CDOing vmlinux.o soon..

**Comment by jpoimboe** [comment]
> At the moment, I am hoping this version will work for our 5.19 based kernel.

If you rely on the ORC unwinder, static calls or return thunks then it may have problems...

**Comment by liu-song-6** [comment]
> > At the moment, I am hoping this version will work for our 5.19 based kernel.
> 
> If you rely on the ORC unwinder, static calls or return thunks then it may have problems...

Hmm... we do use most of these. Let me see how bad it would be.. Maybe we can run objtool on these thinlto files before CDOing?

**Comment by jpoimboe** [comment]
> > > At the moment, I am hoping this version will work for our 5.19 based kernel.
> > 
> > 
> > If you rely on the ORC unwinder, static calls or return thunks then it may have problems...
> 
> Hmm... we do use most of these. Let me see how bad it would be.. Maybe we can run objtool on these thinlto files before CDOing?

Yes, maybe (but to be clear I'd rather avoid this type of Makefile hacking for kpatch upstream)

**Comment by liu-song-6** [comment]
> Yes, maybe (but to be clear I'd rather avoid this type of Makefile hacking for kpatch upstream)

Agreed. CDOing vmlinux.o is a much cleaner approach. 

We already use some non-upstream kpatch commits for PGO. If this hack works for short term, we will do the same.

**Comment by liu-song-6** [comment]
I am testing this change that run objtool on thinlto files: https://github.com/liu-song-6/kpatch/commit/1203c9fc4901c8a53074a36d55166d4cdd037109 . 

I understand we don't want this in upstream, but just in case someone would like to test/use it before we can CDO vmlinux.o.
